File: solutions/python/dnd-character/1/dnd_character.py
import random


class Character:
    def __init__(self):
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom = self.ability()
        self.charisma = self.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        rolls = [random.randint(1, 6) for _ in range(4)]
        return sum(rolls) - min(rolls)


def modifier(value):
    return (value - 10) // 2


if __name__ == '__main__':
    print(Character().__dict__)

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

class DnDCharacter {

    private int strength = ability(rollDice());
    private int dexterity = ability(rollDice());
    private int constitution = ability(rollDice());
    private int intelligence = ability(rollDice());
    private int wisdom = ability(rollDice());
    private int charisma = ability(rollDice());

    int ability(List<Integer> scores) {
        List<Integer> scoresCopy = new ArrayList<>(scores);
        scoresCopy.sort(null);
        return scoresCopy.get(1) + scoresCopy.get(2) + scoresCopy.get(3);
    }

    List<Integer> rollDice() {
        Random random = new Random();
        List<Integer> dice = new ArrayList<>();
        for (int i = 0; i < 4; i++) {
            dice.add(random.nextInt(6) + 1);
        }
        return dice;
    }

    int modifier(int input) {
        return Math.floorDiv(input - 10, 2);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return 10 + modifier(constitution);
    }

}
